### ISSUE TITLE
tf2onnx.convert: invalid attribute data type error

### ISSUE THREAD

**Issue by bluemonk482** [onnx/tensorflow-onnx#184]
Hi, I want to convert my tree-cnn model from tensorflow to onnx, but I am getting error for invalid attribute data type. I am using python 3 btw on Ubuntu 16.04. The `inputs` and `outputs` are identified from using `summarize_graph`:

```
aaa@aaa:~/Tools/tensorflow$ bazel-bin/tensorflow/tools/graph_transforms/summarize_graph --in_graph=../frozen_model.pb
Found 2 possible inputs: (name=inputs/tree, type=float(1), shape=[?,?,183]) (name=inputs/children, type=int32(3), shape=[?,?,?])
No variables spotted.
Found 1 possible outputs: (name=output/Softmax, op=Softmax)
Found 55330 (55.33k) const parameters, 0 (0) variable parameters, and 0 control_edges
Op types used: 124 Const, 19 Pack, 17 Shape, 15 StridedSlice, 10 Fill, 9 ConcatV2, 7 Cast, 7 Mul, 7 Reshape, 7 Identity, 6 Add, 5 Sub, 5 ExpandDims, 4 Tile, 3 Range, 2 Tanh, 2 Prod, 2 Placeholder, 2 Transpose, 2 Minimum, 2 MatMul, 2 GatherV2, 1 NotEqual, 1 Max, 1 RealDiv, 1 ListDiff, 1 Select, 1 Less, 1 Softmax, 1 GreaterEqual, 1 Sum, 1 GatherNd, 1 Equal, 1 BatchMatMul
```

```
aaa@aaa:~/tensorflow-onnx$ python -m tf2onnx.convert --input ../frozen_model.pb --output ../model.onnx --inputs inputs/tree,inputs/children --outputs output/Softmax
/anaconda/envs/py35/lib/python3.5/site-packages/h5py/__init__.py:36: FutureWarning: Conversion of the second argument of issubdtype from `float` to `np.floating` is deprecated. In future, it will be treated as `np.float64 == np.dtype(float).type`.
  from ._conv import register_converters as _register_converters
using tensorflow=1.10.1, onnx=1.2.3, opset=7
2018-11-06 19:11:52.730667: I tensorflow/tools/graph_transforms/transform_graph.cc:318] Applying fold_batch_norms
2018-11-06 19:11:52.735193: I tensorflow/tools/graph_transforms/transform_graph.cc:318] Applying fold_old_batch_norms
2018-11-06 19:11:52.954031: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1405] Found device 0 with properties:
name: Tesla K80 major: 3 minor: 7 memoryClockRate(GHz): 0.8235
pciBusID: b222:00:00.0
totalMemory: 11.17GiB freeMemory: 11.10GiB
2018-11-06 19:11:52.954092: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1484] Adding visible gpu devices: 0
2018-11-06 19:11:53.261597: I tensorflow/core/common_runtime/gpu/gpu_device.cc:965] Device interconnect StreamExecutor with strength 1 edge matrix:
2018-11-06 19:11:53.261646: I tensorflow/core/common_runtime/gpu/gpu_device.cc:971]      0
2018-11-06 19:11:53.261655: I tensorflow/core/common_runtime/gpu/gpu_device.cc:984] 0:   N
2018-11-06 19:11:53.261900: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1097] Created TensorFlow device (/job:localhost/replica:0/task:0/device:GPU:0 with 10758 MB memory) -> physical GPU (device: 0, name: Tesla K80, pci bus id: b222:00:00.0, compute capability: 3.7)
ERROR:tf2onnx:pass1 convert failed for name: "network/conv_layer/conv_node/conv_step/combine/Tensordot/ListDiff"
op: "ListDiff"
input: "network/conv_layer/conv_node/conv_step/combine/Tensordot/range"
input: "network/conv_layer/conv_node/conv_step/combine/Tensordot/add_1"
attr {
  key: "T"
  value {
    type: DT_INT32
  }
}
attr {
  key: "out_idx"
  value {
    type: DT_INT32
  }
}
, ex=Value "<dtype: 'int32'>" is not valid attribute data type.
Traceback (most recent call last):
  File "/anaconda/envs/py35/lib/python3.5/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "/anaconda/envs/py35/lib/python3.5/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/data/home/ccrmad/Tools/tensorflow-onnx/tf2onnx/convert.py", line 118, in <module>
    main()
  File "/data/home/ccrmad/Tools/tensorflow-onnx/tf2onnx/convert.py", line 103, in main
    shape_override=args.shape_override)
  File "/data/home/ccrmad/Tools/tensorflow-onnx/tf2onnx/tfonnx.py", line 1827, in process_tf_graph
    onnx_nodes, op_cnt, attr_cnt, output_shapes, dtypes = tensorflow_to_onnx(tf_graph, shape_override)
  File "/data/home/ccrmad/Tools/tensorflow-onnx/tf2onnx/tfonnx.py", line 130, in tensorflow_to_onnx
    return tflist_to_onnx(graph.get_operations(), shape_override)
  File "/data/home/ccrmad/Tools/tensorflow-onnx/tf2onnx/tfonnx.py", line 117, in tflist_to_onnx
    onnx_node = helper.make_node(node.type, input_names, output_names, name=node.name, **attr)
  File "/anaconda/envs/py35/lib/python3.5/site-packages/onnx/helper.py", line 51, in make_node
    for key, value in sorted(kwargs.items()))
  File "/anaconda/envs/py35/lib/python3.5/site-packages/onnx/helper.py", line 51, in <genexpr>
    for key, value in sorted(kwargs.items()))
  File "/anaconda/envs/py35/lib/python3.5/site-packages/onnx/helper.py", line 234, in make_attribute
    'Value "{}" is not valid attribute data type.'.format(value))
ValueError: Value "<dtype: 'int32'>" is not valid attribute data type.
```

**Comment by guschmue** [comment]
Yes, the ListDiff op has no equivalent in onnx.
Thinking if we can compose this out of other ops.

**Comment by guschmue** [comment]
cool. have some idea how to go about it ? I could not come up with something I like.

**Comment by pengwa** [comment]
@guschmue I found the op is introduced by attention. So it looks that one of its input is a const, if in this case, I guess Gather can do the trick.

**Comment by pengwa** [comment]
@guschmue sorry, Gather won't work. will think about other approach.

**Comment by pengwa** [comment]
@bluemonk482   can you have a try adding "--fold_const", to see whether it works? I assume it will most likely work.

python -m tf2onnx.convert --input ../frozen_model.pb --output ../model.onnx --inputs inputs/tree,inputs/children --outputs output/SoftMax **--fold_const**

**Comment by bluemonk482** [comment]
Thanks for your reply @pengwa ! I have added `--fold_const` however it seems the tensorflow op Range is not supported?

```
$ python -m tf2onnx.convert --input ../frozen_model.pb --output ../model.onnx --inputs inputs/tree,inputs/children --outputs output/Softmax **--fold_const**
using tensorflow=1.12.0-rc0, onnx=1.3.0, opset=7
WARNING:tensorflow:From /anaconda/envs/py35/lib/python3.5/site-packages/tf2onnx-0.0.3.0-py3.5.egg/tf2onnx/convert.py:88: FastGFile.__init__ (from tensorflow.python.platform.gfile) is deprecated and will be removed in a future version.
Instructions for updating:
Use tf.gfile.GFile.
WARNING:tensorflow:From /anaconda/envs/py35/lib/python3.5/site-packages/tf2onnx-0.0.3.0-py3.5.egg/tf2onnx/convert.py:88: FastGFile.__init__ (from tensorflow.python.platform.gfile) is deprecated and will be removed in a future version.
Instructions for updating:
Use tf.gfile.GFile.
WARNING:tensorflow:From /anaconda/envs/py35/lib/python3.5/site-packages/tf2onnx-0.0.3.0-py3.5.egg/tf2onnx/tfonnx.py:1787: extract_sub_graph (from tensorflow.python.framework.graph_util_impl) is deprecated and will be removed in a future version.
Instructions for updating:
Use tf.compat.v1.graph_util.extract_sub_graph
WARNING:tensorflow:From /anaconda/envs/py35/lib/python3.5/site-packages/tf2onnx-0.0.3.0-py3.5.egg/tf2onnx/tfonnx.py:1787: extract_sub_graph (from tensorflow.python.framework.graph_util_impl) is deprecated and will be removed in a future version.
Instructions for updating:
Use tf.compat.v1.graph_util.extract_sub_graph
2018-11-30 20:00:46.449142: I tensorflow/tools/graph_transforms/transform_graph.cc:317] Applying fold_constants
2018-11-30 20:00:46.474642: I tensorflow/tools/graph_transforms/transform_graph.cc:317] Applying remove_attribute
2018-11-30 20:00:46.475744: I tensorflow/tools/graph_transforms/transform_graph.cc:317] Applying fold_batch_norms
2018-11-30 20:00:46.478682: I tensorflow/tools/graph_transforms/transform_graph.cc:317] Applying fold_old_batch_norms
2018-11-30 20:00:46.663770: I tensorflow/compiler/xla/service/service.cc:149] XLA service 0x55d0bdbee8a0 executing computations on platform CUDA. Devices:
2018-11-30 20:00:46.663816: I tensorflow/compiler/xla/service/service.cc:157]   StreamExecutor device (0): Tesla K80, Compute Capability 3.7
2018-11-30 20:00:46.664135: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1432] Found device 0 with properties:
name: Tesla K80 major: 3 minor: 7 memoryClockRate(GHz): 0.8235
pciBusID: 6a0c:00:00.0
totalMemory: 11.17GiB freeMemory: 11.10GiB
2018-11-30 20:00:46.664167: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1511] Adding visible gpu devices: 0
2018-11-30 20:00:46.953192: I tensorflow/core/common_runtime/gpu/gpu_device.cc:982] Device interconnect StreamExecutor with strength 1 edge matrix:
2018-11-30 20:00:46.953251: I tensorflow/core/common_runtime/gpu/gpu_device.cc:988]      0
2018-11-30 20:00:46.953273: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1001] 0:   N
2018-11-30 20:00:46.953523: I tensorflow/core/common_runtime/gpu/gpu_device.cc:1115] Created TensorFlow device (/job:localhost/replica:0/task:0/device:GPU:0 with 10756 MB memory) -> physical GPU (device: 0, name: Tesla K80, pci bus id: 6a0c:00:00.0, compute capability: 3.7)
Traceback (most recent call last):
  File "/anaconda/envs/py35/lib/python3.5/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "/anaconda/envs/py35/lib/python3.5/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/anaconda/envs/py35/lib/python3.5/site-packages/tf2onnx-0.0.3.0-py3.5.egg/tf2onnx/convert.py", line 118, in <module>
  File "/anaconda/envs/py35/lib/python3.5/site-packages/tf2onnx-0.0.3.0-py3.5.egg/tf2onnx/convert.py", line 103, in main
  File "/anaconda/envs/py35/lib/python3.5/site-packages/tf2onnx-0.0.3.0-py3.5.egg/tf2onnx/tfonnx.py", line 1849, in process_tf_graph
  File "/anaconda/envs/py35/lib/python3.5/site-packages/tf2onnx-0.0.3.0-py3.5.egg/tf2onnx/tfonnx.py", line 1720, in tensorflow_onnx_mapping
**ValueError: tensorflow op Range is not supported**
```

**Comment by pengwa** [comment]
ok, range is just supported with PR https://github.com/onnx/tensorflow-onnx/pull/217. not sure whether this is the last issue you will have. let me know if still have further issues.

**Comment by bluemonk482** [comment]
Thanks @pengwa ! The tf op `Range` is now supported but you're right, the new issue now is the support for `FatherNd`:

```
Traceback (most recent call last):
  File "/anaconda/envs/py35/lib/python3.5/runpy.py", line 193, in _run_module_as_main
    "__main__", mod_spec)
  File "/anaconda/envs/py35/lib/python3.5/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/anaconda/envs/py35/lib/python3.5/site-packages/tf2onnx-0.4.0-py3.5.egg/tf2onnx/convert.py", line 117, in <module>
  File "/anaconda/envs/py35/lib/python3.5/site-packages/tf2onnx-0.4.0-py3.5.egg/tf2onnx/convert.py", line 101, in main
  File "/anaconda/envs/py35/lib/python3.5/site-packages/tf2onnx-0.4.0-py3.5.egg/tf2onnx/tfonnx.py", line 2283, in process_tf_graph
  File "/anaconda/envs/py35/lib/python3.5/site-packages/tf2onnx-0.4.0-py3.5.egg/tf2onnx/tfonnx.py", line 2145, in tensorflow_onnx_mapping
ValueError: tensorflow op GatherNd is not supported
```

**Comment by pengwa** [comment]
thanks for this. I guess RNN stuff is good to convert, you can try add "--continue_on_error" to see all unsupported ops in the model. then we can work on them in parallel.

**Comment by bluemonk482** [comment]
Yes, by adding `--continue_on_error` it now creates a `model.onnx` as output. Thanks @pengwa ! I will go and check whether this onnx output works or not. Thanks again.

```
$ python -m tf2onnx.convert --input ../frozen_model.pb --output ../model.onnx --inputs inputs/tree,inputs/children --outputs output/Softmax --fold_const --continue_on_error
using tensorflow=1.12.0-rc0, onnx=1.3.0, opset=7, tfonnx=0.4.0/ba143c
INFO:tf2onnx.optimizer.transpose_optimizer: 1 transpose op(s) left, ops diff after transpose optimization: Counter({'Div': 0, 'Equal': 0, 'Not': 0, 'Mul': 0, 'Sub': 0, 'Add': 0, 'Gather': 0, 'Ceil': 0, 'ReduceSum': 0, 'Tile': 0, 'Slice': 0, 'Identity': 0, 'Softmax': 0, 'Shape': 0, 'Min': 0, 'Concat': 0, 'Squeeze': 0, 'Loop': 0, 'Tanh': 0, 'Reshape': 0, 'ReduceMax': 0, 'ReduceProd': 0, 'Select': 0, 'Unsqueeze': 0, 'MatMul': 0, 'Cast': 0, 'GatherNd': 0, 'Transpose': -1})
```

**Comment by pengwa** [comment]
ok， it should not work, gathernd is not supported. Adding —verbose will provided all mapped and unmapped operators, which would be helpful to understand how far we are to make it work(besides gathernd)

**Comment by bluemonk482** [comment]
Thanks @pengwa ! Here is the mapped and unmapped operators:

```
tensorflow ops: Counter({'Const': 116, 'Pack': 18, 'Shape': 17, 'StridedSlice': 15, 'Fill': 10, 'ConcatV2': 8, 'Reshape': 6, 'Cast': 5, 'ExpandDims': 5, 'Mul': 5, 'Sub': 5, 'Add': 4, 'Tile': 4, 'MatMul': 2, 'Range': 2, 'Prod': 2, 'Placeholder': 2, 'Minimum': 2, 'Tanh': 2, 'GatherV2': 2, 'Transpose': 1, 'Max': 1, 'GatherNd': 1, 'Identity': 1, 'Select': 1, 'Sum': 1, 'Softmax': 1, 'RealDiv': 1, 'NotEqual': 1, 'Equal': 1, 'BatchMatMul': 1})
tensorflow attr: Counter({'dtype': 118, 'value': 116, 'T': 115, 'N': 26, 'axis': 18, 'out_type': 17, 'end_mask': 15, 'begin_mask': 15, 'Index': 15, 'ellipsis_mask': 15, 'shrink_axis_mask': 15, 'new_axis_mask': 15, 'Tidx': 14, 'index_type': 10, 'Tshape': 6, 'SrcT': 5, 'DstT': 5, 'Tdim': 5, 'Truncate': 5, 'keep_dims': 4, 'Tmultiples': 4, 'Tparams': 3, 'Tindices': 3, 'Taxis': 2, 'transpose_b': 2, 'transpose_a': 2, 'shape': 2, 'adj_x': 1, 'adj_y': 1, 'Tperm': 1})
onnx mapped: Counter({'Const': 116, 'Pack': 18, 'Shape': 17, 'StridedSlice': 15, 'Fill': 10, 'ConcatV2': 8, 'Reshape': 6, 'ExpandDims': 5, 'Mul': 5, 'Sub': 5, 'Cast': 5, 'Add': 4, 'Tile': 4, 'MatMul': 2, 'Range': 2, 'Prod': 2, 'Placeholder': 2, 'Minimum': 2, 'Tanh': 2, 'GatherV2': 2, 'Transpose': 1, 'BatchMatMul': 1, 'Equal': 1, 'Max': 1, 'Identity': 1, 'Sum': 1, 'Softmax': 1, 'RealDiv': 1, 'NotEqual': 1})
onnx unmapped: Counter({'Select': 1, 'GatherNd': 1})
INFO:tf2onnx.optimizer.transpose_optimizer: 1 transpose op(s) left, ops diff after transpose optimization: Counter({'MatMul': 0, 'Min': 0, 'ReduceMax': 0, 'Add': 0, 'Slice': 0, 'Cast': 0, 'Ceil': 0, 'ReduceSum': 0, 'Shape': 0, 'Identity': 0, 'Softmax': 0, 'ReduceProd': 0, 'Tile': 0, 'Loop': 0, 'Select': 0, 'Unsqueeze': 0, 'Gather': 0, 'Mul': 0, 'Sub': 0, 'Tanh': 0, 'Concat': 0, 'GatherNd': 0, 'Squeeze': 0, 'Reshape': 0, 'Equal': 0, 'Not': 0, 'Div': 0, 'Transpose': -1})
```

**Comment by pengwa** [comment]
onnx unmapped: Counter({'Select': 1, 'GatherNd': 1})

so GatherNd should be last gap from perspective of op converage. Select is good if you pass --opset=8, but I am actively working on a bug fix now.

**Comment by bluemonk482** [comment]
Thanks @pengwa . Is `tf2onnx-0.4.0` the latest version? 

By using the the following command I am getting the AttributeError:

`python -m tf2onnx.convert --input /frozen_model.pb --output /model.onnx --inputs inputs/tree,inputs/children --outputs output/Softmax --fold_const`


> ERROR:tf2onnx:node <function rewrite_custom_rnn_cell at 0x7f8d1fb42a60>: exception 'NoneType' object has no attribute 'get_implicit_inputs'
> Traceback (most recent call last):
>   File "/anaconda/envs/py35/lib/python3.5/runpy.py", line 193, in _run_module_as_main
>     "__main__", mod_spec)
>   File "/anaconda/envs/py35/lib/python3.5/runpy.py", line 85, in _run_code
>     exec(code, run_globals)
>   File "/anaconda/envs/py35/lib/python3.5/site-packages/tf2onnx-0.4.0-py3.5.egg/tf2onnx/convert.py", line 121, in <module>
>   File "/anaconda/envs/py35/lib/python3.5/site-packages/tf2onnx-0.4.0-py3.5.egg/tf2onnx/convert.py", line 102, in main
>   File "/anaconda/envs/py35/lib/python3.5/site-packages/tf2onnx-0.4.0-py3.5.egg/tf2onnx/tfonnx.py", line 2395, in process_tf_graph
>   File "/anaconda/envs/py35/lib/python3.5/site-packages/tf2onnx-0.4.0-py3.5.egg/tf2onnx/tfonnx.py", line 2386, in process_tf_graph
>   File "/anaconda/envs/py35/lib/python3.5/site-packages/tf2onnx-0.4.0-py3.5.egg/tf2onnx/rewriter/rnn.py", line 50, in rewrite_custom_rnn_cell
>   File "/anaconda/envs/py35/lib/python3.5/site-packages/tf2onnx-0.4.0-py3.5.egg/tf2onnx/rewriter/custom_rnn_rewriter.py", line 78, in run
>   File "/anaconda/envs/py35/lib/python3.5/site-packages/tf2onnx-0.4.0-py3.5.egg/tf2onnx/rewriter/loop_rewriter_base.py", line 76, in run_internal
>   File "/anaconda/envs/py35/lib/python3.5/site-packages/tf2onnx-0.4.0-py3.5.egg/tf2onnx/graph.py", line 940, in delete_unused_nodes
>   File "/anaconda/envs/py35/lib/python3.5/site-packages/tf2onnx-0.4.0-py3.5.egg/tf2onnx/graph.py", line 925, in extract_sub_graph_nodes
>   File "/anaconda/envs/py35/lib/python3.5/site-packages/tf2onnx-0.4.0-py3.5.egg/tf2onnx/graph.py", line 908, in _extract_sub_graph_nodes
> AttributeError: 'NoneType' object has no attribute 'get_implicit_inputs'

**Comment by pengwa** [comment]
Oh, yeah, the change is not included in latest release. you can use master. and in the command line, you have to specify --opset 8, and inputs should be "inputs/tree:0,inputs/children:0", outputs should be "--outputs output/SoftMax:0"

**Comment by mnboos** [comment]
I have this issue too:

```
ERROR:tf2onnx:pass1 convert failed for name: "gradients/bidirectional_rnn/bw/bw/while/TensorArrayWrite/TensorArrayWriteV3_grad/TensorArrayReadV3/f_acc"
op: "StackV2"
input: "gradients/bidirectional_rnn/bw/bw/while/TensorArrayWrite/TensorArrayWriteV3_grad/TensorArrayReadV3/Const"
attr {
  key: "elem_type"
  value {
    type: DT_INT32
  }
}
attr {
  key: "stack_name"
  value {
    s: ""
  }
}
, ex=Value "<dtype: 'int32'>" is not valid attribute data type.
Traceback (most recent call last):
  File "/usr/lib/python3.5/runpy.py", line 184, in _run_module_as_main
    "__main__", mod_spec)
  File "/usr/lib/python3.5/runpy.py", line 85, in _run_code
    exec(code, run_globals)
  File "/usr/local/lib/python3.5/dist-packages/tf2onnx-0.4.0-py3.5.egg/tf2onnx/convert.py", line 123, in <module>
  File "/usr/local/lib/python3.5/dist-packages/tf2onnx-0.4.0-py3.5.egg/tf2onnx/convert.py", line 105, in main
  File "/usr/local/lib/python3.5/dist-packages/tf2onnx-0.4.0-py3.5.egg/tf2onnx/tfonnx.py", line 2472, in process_tf_graph
  File "/usr/local/lib/python3.5/dist-packages/tf2onnx-0.4.0-py3.5.egg/tf2onnx/tfonnx.py", line 144, in tensorflow_to_onnx
  File "/usr/local/lib/python3.5/dist-packages/tf2onnx-0.4.0-py3.5.egg/tf2onnx/tfonnx.py", line 131, in tflist_to_onnx
  File "/usr/local/lib/python3.5/dist-packages/onnx-1.4.1-py3.5-linux-x86_64.egg/onnx/helper.py", line 56, in make_node
    for key, value in sorted(kwargs.items()))
  File "/usr/local/lib/python3.5/dist-packages/onnx-1.4.1-py3.5-linux-x86_64.egg/onnx/helper.py", line 56, in <genexpr>
    for key, value in sorted(kwargs.items()))
  File "/usr/local/lib/python3.5/dist-packages/onnx-1.4.1-py3.5-linux-x86_64.egg/onnx/helper.py", line 255, in make_attribute
    'Value "{}" is not valid attribute data type.'.format(value))
ValueError: Value "<dtype: 'int32'>" is not valid attribute data type.
```

**Comment by mnboos** [comment]
I opened a new issue: https://github.com/onnx/tensorflow-onnx/issues/184
